src/mesh.cpp

```diff
@@ -0,0 +1,981 @@
+        "the number of entries on <lower_left>.");
+    }
+
+    // Check for negative widths
+    if (xt::any(width_ < 0.0)) {
+      fatal_error("Cannot have a negative <width> on a tally mesh.");
+    }
+
+    // Set width and upper right coordinate
+    upper_right_ = xt::eval(lower_left_ + shape_ * width_);
+
+  } else if (check_for_node(node, "upper_right")) {
+    upper_right_ = get_node_xarray<double>(node, "upper_right");
+
+    // Check to ensure width has same dimensions
+    auto n = upper_right_.size();
+    if (n != lower_left_.size()) {
+      fatal_error("Number of entries on <upper_right> must be the "
+        "same as the number of entries on <lower_left>.");
+    }
+
+    // Check that upper-right is above lower-left
+    if (xt::any(upper_right_ < lower_left_)) {
+      fatal_error("The <upper_right> coordinates must be greater than "
+        "the <lower_left> coordinates on a tally mesh.");
+    }
+
+    // Set width and upper right coordinate
+    width_ = xt::eval((upper_right_ - lower_left_) / shape_);
+  } else {
+    fatal_error("Must specify either <upper_right> and <width> on a mesh.");
+  }
+
+  if (shape_.dimension() > 0) {
+    if (shape_.size() != lower_left_.size()) {
+      fatal_error("Number of entries on <lower_left> must be the same "
+        "as the number of entries on <dimension>.");
+    }
+
+    // Set volume fraction
+    volume_frac_ = 1.0/xt::prod(shape_)();
+  }
+}
+
+int RegularMesh::get_bin(Position r)
+{
+  // Loop over the dimensions of the mesh
+  for (int i = 0; i < n_dimension_; ++i) {
+    // Check for cases where particle is outside of mesh
+    if (r[i] < lower_left_[i]) {
+      return -1;
+    } else if (r[i] > upper_right_[i]) {
+      return -1;
+    }
+  }
+
+  // Determine indices
+  int ijk[n_dimension_];
+  bool in_mesh;
+  get_indices(r, ijk, &in_mesh);
+  if (!in_mesh) return -1;
+
+  // Convert indices to bin
+  return get_bin_from_indices(ijk);
+}
+
+int RegularMesh::get_bin_from_indices(const int* ijk)
+{
+  if (n_dimension_ == 1) {
+    return ijk[0];
+  } else if (n_dimension_ == 2) {
+    return (ijk[1] - 1)*shape_[0] + ijk[0];
+  } else if (n_dimension_ == 3) {
+    return ((ijk[2] - 1)*shape_[1] + (ijk[1] - 1))*shape_[0] + ijk[0];
+  }
+}
+
+void RegularMesh::get_indices(Position r, int* ijk, bool* in_mesh)
+{
+  // Find particle in mesh
+  *in_mesh = true;
+  for (int i = 0; i < n_dimension_; ++i) {
+    ijk[i] = std::ceil((r[i] - lower_left_[i]) / width_[i]);
+
+    // Check if indices are within bounds
+    if (ijk[i] < 1 || ijk[i] > shape_[i]) *in_mesh = false;
+  }
+}
+
+void RegularMesh::get_indices_from_bin(int bin, int* ijk)
+{
+  if (n_dimension_ == 1) {
+    ijk[0] = bin;
+  } else if (n_dimension_ == 2) {
+    ijk[0] = (bin - 1) % shape_[0] + 1;
+    ijk[1] = (bin - 1) / shape_[0] + 1;
+  } else if (n_dimension_ == 3) {
+    ijk[0] = (bin - 1) % shape_[0] + 1;
+    ijk[1] = (bin - 1) % (shape_[0] * shape_[1]) / shape_[0] + 1;
```

Comment:
There are different conventions for modulo in the order of operations.  Stick some parentheses around it just to remove any possible ambiguity for future readers.

Reply:
Good call, `%` and `/` are at the same operator precedence level so this only works because of left-to-right guarantee. I'll add an extra set of parentheses to remove any ambiguity.